Mathlib/Analysis/InnerProductSpace/TwoDim.lean

```diff
@@ -0,0 +1,676 @@
+  rintro (rfl | rfl) <;> simp
+#align orientation.kahler_eq_zero_iff Orientation.kahler_eq_zero_iff
+
+theorem kahler_ne_zero {x y : E} (hx : x ≠ 0) (hy : y ≠ 0) : o.kahler x y ≠ 0 := by
+  apply mt o.eq_zero_or_eq_zero_of_kahler_eq_zero
+  tauto
+#align orientation.kahler_ne_zero Orientation.kahler_ne_zero
+
+theorem kahler_ne_zero_iff (x y : E) : o.kahler x y ≠ 0 ↔ x ≠ 0 ∧ y ≠ 0 := by
+  refine' ⟨_, fun h => o.kahler_ne_zero h.1 h.2⟩
+  contrapose
+  simp only [not_and_or, Classical.not_not, kahler_apply_apply, Complex.real_smul]
+  rintro (rfl | rfl) <;> simp
+#align orientation.kahler_ne_zero_iff Orientation.kahler_ne_zero_iff
+
+theorem kahler_map {F : Type _} [NormedAddCommGroup F] [InnerProductSpace ℝ F]
+    [Fact (finrank ℝ F = 2)] (φ : E ≃ₗᵢ[ℝ] F) (x y : F) :
+    (Orientation.map (Fin 2) φ.toLinearEquiv o).kahler x y = o.kahler (φ.symm x) (φ.symm y) := by
+  simp [kahler_apply_apply, areaForm_map]
+#align orientation.kahler_map Orientation.kahler_map
+
+/-- The bilinear map `kahler` is invariant under pullback by a positively-oriented isometric
+automorphism. -/
+theorem kahler_comp_linearIsometryEquiv (φ : E ≃ₗᵢ[ℝ] E)
+    (hφ : 0 < LinearMap.det (φ.toLinearEquiv : E →ₗ[ℝ] E)) (x y : E) :
+    o.kahler (φ x) (φ y) = o.kahler x y :=
+  by simp [kahler_apply_apply, o.areaForm_comp_linearIsometryEquiv φ hφ]
+#align orientation.kahler_comp_linear_isometry_equiv Orientation.kahler_comp_linearIsometryEquiv
+
+end Orientation
+
+namespace Complex
+
+attribute [local instance] Complex.finrank_real_complex_fact
+
+@[simp]
+protected theorem areaForm (w z : ℂ) : Complex.orientation.areaForm w z = (conj w * z).im := by
+  let o := Complex.orientation
+  -- Porting note: split `simp only` for greater proof control
+  rw [o.areaForm_to_volumeForm, o.volumeForm_robust Complex.orthonormalBasisOneI rfl,
+    Basis.det_apply, Matrix.det_fin_two]
+  repeat rw [Basis.toMatrix_apply]
+  simp only [toBasis_orthonormalBasisOneI, Matrix.cons_val_zero, Matrix.cons_val_one,
+    Matrix.head_cons, coe_basisOneI_repr, mul_im, conj_re, conj_im, neg_mul]
+  ring
+#align complex.area_form Complex.areaForm
+
+@[simp]
+protected theorem rightAngleRotation (z : ℂ) :
+    Complex.orientation.rightAngleRotation z = I * z := by
+  apply ext_inner_right ℝ
+  intro w
+  rw [Orientation.inner_rightAngleRotation_left]
+  simp only [Complex.areaForm, Complex.inner, mul_re, mul_im, conj_re, conj_im, map_mul, conj_I,
+    neg_re, neg_im, I_re, I_im]
+  ring
+#align complex.right_angle_rotation Complex.rightAngleRotation
+
+@[simp]
+protected theorem kahler (w z : ℂ) : Complex.orientation.kahler w z = conj w * z := by
+  rw [Orientation.kahler_apply_apply]
+  ext1 <;> simp
+#align complex.kahler Complex.kahler
+
+end Complex
+
+namespace Orientation
+
+local notation "ω" => o.areaForm
+
+local notation "J" => o.rightAngleRotation
+
+open Complex
+
+attribute [local instance] finrank_real_complex_fact -- Porting note: this instance is needed
```

Comment:
Was it not here in mathlib3? What has changed?

Reply:
It wasn't in mathlib3, but without this in mathlib4 the three following theorems have the error
```
failed to synthesize instance
  Fact (finrank ℝ ℂ = 2)
```

Reply:
The alternative (as I have now implemented) is to name the `[Fact (finrank ℝ F = 2)]` arguments of the relevant declarations – I used `hF` – and then put `(hF := _)`.

Reply:
Please put the explanation in the porting note if possible!

Reply:
@semorrison Done.